### PR TITLE
Fix #10758 Add compiler option to parse in strict mode

### LINKED ISSUE #10758: Add compiler option to parse in strict mode
See #9449 and #10755 for reference.

The ts code base is not modules, so it is not implicitly parsed in strict mode. we would like it to be. options either 1. put "use strict" at the top of every file, and make the emitter remove the duplicate ones when it merges the output. or 2. add a compiler flag that would force the compiler to parse the files in strict mode.

### PR DESCRIPTION
**Pull request by slawomir**
Fixes #10758

 * add compiler option alwaysStrict
 * compile in strict mode when option is set
 * emit "use strict"

### CODE REVIEW

File: tests/baselines/reference/alwaysStrictNoImplicitUseStrict.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+//// [alwaysStrictNoImplicitUseStrict.ts]
+
+module M {
+    export function f() {
+        var arguments = [];
+    }
+}
+
+//// [alwaysStrictNoImplicitUseStrict.js]
+"use strict";
```
Comment (slawomir) [inline tests/baselines/reference/alwaysStrictNoImplicitUseStrict.js]: alwaysStrict overrides noImplicitUseStrict

File: src/compiler/factory.ts [hunk 2]
```diff
@@ -2236,6 +2236,36 @@ namespace ts {
...
     /**
+     * Ensures "use strict" directive is added
+     * 
+     * @param node source file
+     */
+    export function ensureUseStrict(node: SourceFile): SourceFile {
+        let foundUseStrict = false;
+        let statementOffset = 0;
+        const numStatements = node.statements.length;
+        while (statementOffset < numStatements) {
```
Comment (RyanCavanaugh) [inline src/compiler/factory.ts]: Why isn't this a `for` or `for / of` loop?

Comment (RyanCavanaugh) [inline src/compiler/factory.ts]: Seems like you were planning to use `statementOffset` later and then didn't

File: src/compiler/transformers/ts.ts [hunk 3]
```diff
@@ -436,6 +436,11 @@ namespace ts {
         function visitSourceFile(node: SourceFile) {
             currentSourceFile = node;
 
+            // ensure "use strict"" is emitted in all scenarios in alwaysStrict mode
```
Comment (RyanCavanaugh) [inline src/compiler/transformers/ts.ts]: nit: extra quote

**Comment by slawomir** [comment]
@RyanCavanaugh thanks for analysing. I will send next iteration.

File: src/compiler/transformers/ts.ts (hunk 3 above)
Comment (slawomir) [inline src/compiler/transformers/ts.ts]: Fixed

File: tests/baselines/reference/alwaysStrictNoImplicitUseStrict.js (hunk 1 above)
Comment (mhegazy) [inline tests/baselines/reference/alwaysStrictNoImplicitUseStrict.js]: Can you add a check in program.ts::verifyCompilerOptions to flag teh use of `--noImplicitUseStrict` and `--alwaysStrict` as invalid combination.. so somthing like
```ts
if (options.noImplicitUseStrict && options.alwaysStrict) {
    programDiagnostics.add(createCompilerDiagnostic(Diagnostics.Option_0_cannot_be_specified_with_option_1, "noImplicitUseStrict ", "alwaysStrict"));
}
```

**Comment by mhegazy** [comment]
Looks good. one comment for flagging invalid combinations.

**Comment by mhegazy** [comment]
also can you add a test for `--outFile` concatenating multiple files.

**Pull request merged**